README.md
# ![Signatory][logo]

[![crate][crate-image]][crate-link]
[![Docs][docs-image]][docs-link]
![Apache2/MIT licensed][license-image]
![MSRV][rustc-image]
[![Build Status][build-image]][build-link]

Pure Rust multi-provider digital signature library with support for elliptic
curve digital signature algorithms, namely ECDSA (described in [FIPS 186‑4])
and Ed25519 (described in [RFC 8032]).

Signatory provides a thread-safe and object-safe API and implements providers
for many popular Rust crates, including [ed25519‑dalek], [*ring*], [secp256k1], 
and [sodiumoxide].

[Documentation][docs-link]

## About

Signatory exposes a thread-and-object-safe API for creating digital signatures
which allows several signature providers to be compiled-in and available with
specific providers selected at runtime.

## Requirements

All Signatory providers require Rust **1.37+**

## Provider Support

Signatory includes the following providers, which are each packaged into their
own respective crates (except for the [yubihsm] provider, which is included
[directly in the yubihsm crate]).

### ECDSA providers

| Provider Crate        | Backend Crate  | Type | P‑256  | P‑384  | secp256k1  |
| --------------------- | -------------- | ---- | ------ | ------ | ---------- |
| [signatory‑ring]      | [*ring*]       | Soft | ✅     | ✅     | ⛔         |
| [signatory‑secp256k1] | [secp256k1]    | Soft | ⛔     | ⛔     | ✅         |
| [yubihsm]             | [yubihsm]      | Hard | ✅     | ✅     | ✅         |

### Ed25519 providers

| Provider Crate          | Backend Crate   | Type | Signing | Verification |
| ----------------------- | --------------- | ---- | ------- | ------------ |
| [signatory‑dalek]       | [ed25519‑dalek] | Soft | 51 k/s  | 18 k/s       |
| [signatory‑ring]        | [*ring*]        | Soft | 47 k/s  | 16 k/s       |
| [signatory‑sodiumoxide] | [sodiumoxide]   | Soft | 38 k/s  | 15 k/s       |
| [yubihsm]               | [yubihsm]       | Hard | ~8/s    | N/A          |

### Tendermint only providers (amino encoded consensus votes)

| Provider Crate        | Backend Crate   | Type | Signing | Verification |
| --------------------- | --------------- | ---- | ------- | ------------ |
| [signatory‑ledger-tm] | [ledger-tendermint] | Hard | N/A     | N/A          |

Above benchmarks performed using `cargo bench` on an Intel Xeon E3-1225 v5 @ 3.30GHz.

## License

**Signatory** is distributed under the terms of either the MIT license or the
Apache License (Version 2.0), at your option.

See [LICENSE-APACHE](LICENSE-APACHE) and [LICENSE-MIT](LICENSE-MIT) for details.

[//]: # (badges)

[logo]: https://storage.googleapis.com/iqlusion-production-web/github/signatory/signatory.svg
[crate-image]: https://img.shields.io/crates/v/signatory.svg
[crate-link]: https://crates.io/crates/signatory
[docs-image]: https://docs.rs/signatory/badge.svg
[docs-link]: https://docs.rs/signatory/
[license-image]: https://img.shields.io/badge/license-Apache2.0/MIT-blue.svg
[rustc-image]: https://img.shields.io/badge/rustc-1.37+-blue.svg
[build-image]: https://github.com/iqlusioninc/signatory/workflows/Rust/badge.svg?branch=develop&event=push
[build-link]: https://github.com/iqlusioninc/signatory/actions

[//]: # (general links)

[FIPS 186‑4]: https://csrc.nist.gov/publications/detail/fips/186/4/final
[RFC 8032]: https://tools.ietf.org/html/rfc8032
[ed25519‑dalek]: https://github.com/dalek-cryptography/ed25519-dalek
[*ring*]: https://github.com/briansmith/ring
[secp256k1]: https://github.com/rust-bitcoin/rust-secp256k1/
[sodiumoxide]: https://github.com/dnaq/sodiumoxide
[yubihsm]: https://github.com/tendermint/yubihsm-rs
[ledger-tendermint]: https://crates.io/crates/ledger-tendermint
[directly in the yubihsm crate]: https://docs.rs/yubihsm/latest/yubihsm/signatory/index.html
[signatory‑dalek]: https://crates.io/crates/signatory-dalek
[signatory‑ring]: https://crates.io/crates/signatory-ring
[signatory‑secp256k1]: https://crates.io/crates/signatory-secp256k1
[signatory‑sodiumoxide]: https://crates.io/crates/signatory-sodiumoxide
[signatory‑ledger-tm]: https://crates.io/crates/signatory-ledger-tm


CHANGES.md
## 0.19.0 (2020-04-19)

- Update `signature` crate requirement to v1.0.1 ([#14])
- Update `subtle-encoding` requirement to 0.5 ([#2])

[#14]: https://github.com/iqlusioninc/signatory/pull/14
[#2]: https://github.com/iqlusioninc/signatory/pull/2

## 0.18.1 (2020-03-02)

- Update links to point to new repository location
- signatory-secp256k1: update `secp256k1` requirement from 0.15 to 0.17
- signatory-secp256k1: support Ethereum's `Keccak256` hash function for signing

## 0.18.0 (2020-01-19)

- Upgrade `ecdsa` crate to v0.4

## 0.17.1 (2019-12-16)

- Fix Windows builds

## 0.17.0 (2019-12-11)

- Use `=` expression to lock all prerelease deps to specific versions
- Upgrade to `secp256k1` crate v0.17
- Upgrade to `ecdsa` crate v0.3

## 0.16.0 (2019-10-29)

- Use the `ecdsa` crate
- Upgrade to zeroize 1.0
- Upgrade to `signature` and `ed25519` crates v1.0.0-pre.1

## 0.15.0 (2019-10-11)

- signatory-dalek: Upgrade `ed25519-dalek` to 1.0.0-pre.2
- Upgrade to `signature` and `ed25519` crates v1.0.0-pre.0

## 0.14.0 (2019-10-10)

- Always use the `alloc` crate for `String`/`Vec`
- Upgrade to `signature` crate v0.3, `ed25519` crate v0.2; MSRV 1.36+

## 0.13.0 (2019-08-11)

- Update ring to v0.16; secp256k1 to v0.15
- Remove toplevel `signature` re-exports; add `encoding::Error`
- Use `ed25519::Signature` from the `ed25519` crate

## 0.12.0 (2019-06-07)

- Use the `signature` crate

## 0.11.5 (2019-06-04)

- Upgrade to `zeroize` 0.9

## 0.11.4 (2019-05-20)

- Support stable `alloc` API
- Upgrade to `zeroize` 0.8

## 0.11.3 (2019-03-13)

- Fix Missing TrailingWhitespace type-case in subtle-encoding error conversion

## 0.11.2 (2019-03-09)

- ecdsa: impl `PartialOrd` + `Ord` for PublicKeys
- ecdsa: Simplify trait bounds for Copy impl on curve point types

## 0.11.1 (2019-02-23)

- ecdsa: impl `Copy` + `Hash` for ECDSA curve points and public keys

## 0.11.0 (2019-02-12)

- signatory-yubihsm: Update `yubihsm` crate to v0.20
- signatory-dalek: Update `ed25519-dalek` crate to 1.0.0-pre.1
- signatory-ring: Update `ring` crate to 0.14
- signatory-sodiumoxide: Update `sodiumoxide` crate to 0.2
- signatory-secp256k1: Update `secp256k1` crate to 0.12
- Upgrade to Rust 2018 edition
- signatory-ledger-cosval: Upgrade ledger provider to validator app 0.2.1

## 0.10.1 (2018-11-27)

- Upgrade to `subtle-encoding` v0.3.0

## 0.10.0 (2018-10-16)

- Upgrade to `digest` 0.8, `generic-array` 0.12, and `yubihsm` 0.19
- Upgrade to `zeroize` 0.4

## 0.9.4 (2018-10-10)

- pkcs8: Properly gate `FILE_MODE` on Windows

## 0.9.3 (2018-10-09)

- Upgrade to `subtle-encoding` v0.2
- Fix unused import on Windows (closes #121)

## 0.9.2 (2018-10-08)

- More documentation fixups

## 0.9.1 (2018-10-08)

- Cargo.toml: Fix docs.rs build

## 0.9.0 (2018-10-08)

- Remove redundant "namespacing" from type names
- Move `curve` module (back) under `ecdsa`
- signatory-yubihsm: Upgrade to yubihsm 0.18
- Use `subtle-encoding` crate for constant-time encoding/decoding
- ECDSA `SecretKey` type and related traits (e.g. `GeneratePkcs8`)
- Properly handle leading zeroes in ASN.1 serialization/parsing
- signatory-yubihsm: Expose the yubihsm crate as a pub extern
- encoding: Use 0o600 file mode on Unix
- Eliminate `ed25519::FromSeed` trait
- yubihsm: NIST P-384 support
- ring: NIST P-384 support
- Add NIST P-384 elliptic curve type (closes #73)
- signatory-yubihsm: Fix ECDSA over secp256k1 signing (closes #87)
- `signatory-ledger-cosval` provider
- signatory-yubihsm: Normalize secp256k1 signatures to "low S" form
- signatory-secp256k1: Bump secp256k1 crate dependency to 0.11
- Unify verification API under the `Verifier` trait
- encoding: Add encoding module with hex and Base64 support
- Unify signing API under the `Signer` trait

## 0.8.0 (2018-08-19)

- Extract `from_pkcs8` into a trait
- signatory-yubihsm: Make ecdsa and ed25519 modules public

## 0.7.0 (2018-08-19)

- Factor providers into their own `signatory-*` crates
- Unify ECDSA traits across DER and fixed-sized signatures
- ECDSA DER signature parsing and serialization

## 0.6.1 (2018-07-31)

- Upgrade to `secp256k1` crate v0.10

## 0.6.0 (2018-07-31)

- Factor ECDSA PublicKey into compressed/uncompressed curve points
- ECDSA support for `yubihsm-provider`
- Upgrade to `yubihsm` crate 0.14
- Add rustdoc logo
- Audit project for security vulnerabilities with `cargo-audit`
- Update to `ed25519-dalek` 0.8
- Add ECDSA NIST P-256 support with *ring* provider
- Factor ECDSA traits apart into separate traits per method
- Upgrade to `sodiumoxide` 0.1
- Add `ed25519::Seed::from_keypair` method
- No default features
- Add `ed25519::Seed` type

## 0.5.2 (2018-05-19)

- Update to `yubihsm-rs` 0.9
- Fix benchmarks

## 0.5.1 (2018-04-13)

- Mark all Signers and Verifiers as Send safe

## 0.5.0 (2018-04-12)

- Upgrade to `yubihsm-rs` 0.8
- ECDSA verification support
- ECDSA support with secp256k1 provider
- Ed25519 FromSeed trait and miscellaneous cleanups
- Remove unnecessary direct dependency on `curve25519-dalek`

## 0.4.1 (2018-04-05)

- Add more bounds to the `Verifier` trait

## 0.4.0 (2018-04-05)

- Add an `ed25519` module to all providers
- `sodiumoxide` provider for Ed25519
- *ring* Ed25519 provider
- `ed25519::Verifier` trait

## 0.3.2 (2018-03-31)

- Upgrade `ed25519-dalek` to 0.6.2

## 0.3.1 (2018-03-27)

- Update to `yubihsm-rs` 0.7

## 0.3.0 (2018-03-20)

- Refactor providers + `yubihsm-rs` update + `Sync`-safe signers

## 0.2.0 (2018-03-13)

- Add `ed25519::Signer::public_key()`

## 0.1.0 (2018-03-12)

- Initial release


src/lib.rs
//! Signatory: a multi-provider digital signature library
//!
//! This crate provides a thread-and-object-safe API for both creating and
//! verifying elliptic curve digital signatures, using either software-based
//! or hardware-based providers.
//!
//! The following algorithms are supported:
//!
//! - [ecdsa]: Elliptic Curve Digital Signature Algorithm ([FIPS 186-4])
//! - [ed25519]: Edwards Digital Signature Algorithm (EdDSA) instantiated using
//!   the twisted Edwards form of Curve25519 ([RFC 8032]).
//!
//! ## Providers
//!
//! There are several backend providers available, which are each available
//! in their own crates:
//!
//! - [signatory-dalek]: Ed25519 signing/verification using the pure-Rust
//!   [ed25519-dalek] crate.
//! - [signatory-ring]: ECDSA and Ed25519 signing/verification provider
//!   for the [*ring*] cryptography library.
//! - [signatory-secp256k1]: ECDSA signing/verification for the secp256k1
//!   elliptic curve (commonly used by Bitcoin and other cryptocurrrencies)
//!   which wraps the [libsecp256k1] library from Bitcoin Core.
//! - [signatory-sodiumoxide]: Ed25519 signing/verification with the
//!   [sodiumoxide] crate, a Rust wrapper for libsodium (NOTE: requires
//!   libsodium to be installed on the system)
//! - [yubihsm-rs]: ECDSA and Ed25519 signing provider support for
//!   private keys stored in a `YubiHSM2` hardware device, via the
//!   Signatory signers types in the [yubihsm-rs] crate
//!   ([yubihsm::ecdsa::Signer] and [yubihsm::ed25519::Signer]).
//!
//! ## Signing API
//!
//! - [Signer]: trait for signing
//! - [DigestSigner]: trait for signing digests
//!
//! ## Verifier API
//!
//! - [Verifier]: trait for verifying
//! - [DigestVerifier]: trait for verifying digests
//!
//! [FIPS 186-4]: https://csrc.nist.gov/publications/detail/fips/186/4/final
//! [RFC 8032]: https://tools.ietf.org/html/rfc8032
//! [ecdsa]: https://docs.rs/signatory/latest/signatory/ecdsa/index.html
//! [ed25519]: https://docs.rs/signatory/latest/signatory/ed25519/index.html
//! [signatory-dalek]: https://docs.rs/crate/signatory-dalek/
//! [ed25519-dalek]: https://docs.rs/crate/ed25519-dalek/
//! [signatory-ring]: https://docs.rs/crate/signatory-ring/
//! [*ring*]: https://github.com/briansmith/ring
//! [signatory-secp256k1]: https://docs.rs/crate/signatory-secp256k1/
//! [libsecp256k1]: https://docs.rs/crate/secp256k1
//! [signatory-sodiumoxide]: https://docs.rs/crate/signatory-sodiumoxide/
//! [sodiumoxide]: https://docs.rs/crate/sodiumoxide/
//! [yubihsm-rs]: https://docs.rs/crate/yubihsm/
//! [yubihsm::ecdsa::Signer]: https://docs.rs/yubihsm/latest/yubihsm/ecdsa/struct.Signer.html
//! [yubihsm::ed25519::Signer]: https://docs.rs/yubihsm/latest/yubihsm/ed25519/struct.Signer.html
//! [Signer]: https://docs.rs/signatory/latest/signatory/trait.Signer.html
//! [DigestSigner]: https://docs.rs/signatory/latest/signatory/trait.DigestSigner.html
//! [Verifier]: https://docs.rs/signatory/latest/signatory/trait.Verifier.html
//! [DigestVerifier]: https://docs.rs/signatory/latest/signatory/trait.DigestVerifier.html

#![no_std]
#![forbid(unsafe_code)]
#![warn(missing_docs, rust_2018_idioms, unused_qualifications)]
#![doc(
    html_logo_url = "https://raw.githubusercontent.com/iqlusioninc/signatory/develop/img/signatory-rustacean.png",
    html_root_url = "https://docs.rs/signatory/0.19.0"
)]

#[cfg(feature = "alloc")]
extern crate alloc;

#[cfg(any(feature = "std", test))]
#[macro_use]
extern crate std;

#[cfg(feature = "ecdsa")]
pub mod ecdsa;
#[cfg(feature = "ed25519")]
#[macro_use]
pub mod ed25519;
#[cfg(feature = "encoding")]
pub mod encoding;
pub mod public_key;
#[cfg(feature = "test-vectors")]
pub mod test_vector;
#[cfg(feature = "generic-array")]
pub use generic_array;
#[cfg(feature = "sha2")]
pub use sha2;
pub use signature;
